Payara can close JarFile instances used by current URLClassLoaders

# Description #
----------

Several places in the Payara code call JarFileUtils.closeCachedJarFiles (Weld deployment, undeployment, etc) in order to clean up no longer needed open files.  The problem is that the code cannot determine why the JarFile is open.  Unfortunately in certain circumstances URLClassLoader can open JarFile objects that are stored in that cache and forcing the closure breaks the class loader for all further class loading attempts.

## Expected Outcome

Classloaders should continue working until the classloader itself is closed.

## Current Outcome

Under certain circumstances (URLClassLoaders with a URL NOT in the "file:" schema), the JarFile cleanup causes the class loader to break.

## Steps to reproduce (Only for bug reports) 

The circumstances we saw this was in using JRuby based servlet applications (Ruby on Rails) where it was using internal JRuby SSL code (in the form of the jopenssl.jar file embedded in the jruby-stdlibs jar).

## Environment ##

- **Payara Version**: 5.194
- **Edition**: Full
- **JDK Version**: Open JDK 8
- **Operating System**: RedHat Linux

I would like to renew this discussion.

I have created a sample set of applications that reproduce the issue caused by cached `JarFile` closures.

[payara-repro-app.zip](https://github.com/payara/Payara/files/5363783/payara-repro-app.zip)

To reproduce the issue, build the project with Maven, deploy `crashing-app`, make a request against it and then deploy `app-that-causes-crashing`. 

One should get the following exception:
```
[#|2020-10-12T10:36:02.815+0300|WARNING|Payara 5.2020|javax.enterprise.web|_ThreadID=102;_ThreadName=http-thread-pool::http-listener-1(4);_TimeMillis=1602488162815;_LevelValue=900;|
  StandardWrapperValve[com.test.CrashingServlet]: Servlet.service() for servlet com.test.CrashingServlet threw exception
java.lang.RuntimeException: java.lang.IllegalStateException: zip file closed
	at com.test.CrashingServlet.doGet(CrashingServlet.java:30)
...
Caused by: java.lang.IllegalStateException: zip file closed
	at java.util.zip.ZipFile.ensureOpen(ZipFile.java:686)
	at java.util.zip.ZipFile.access$200(ZipFile.java:60)
	at java.util.zip.ZipFile$ZipEntryIterator.next(ZipFile.java:519)
	at java.util.zip.ZipFile$ZipEntryIterator.nextElement(ZipFile.java:514)
	at java.util.zip.ZipFile$ZipEntryIterator.nextElement(ZipFile.java:495)
	at java.util.jar.JarFile$JarEntryIterator.next(JarFile.java:257)
	at java.util.jar.JarFile$JarEntryIterator.nextElement(JarFile.java:266)
	at java.util.jar.JarFile$JarEntryIterator.nextElement(JarFile.java:247)
	at com.test.CrashingServlet.doGet(CrashingServlet.java:24)
	... 30 more
```

The repro application is artificial, but hopefully it proves the point that the deployment of an application `B` can be destructive for an unrelated application `A`.

I stumbled upon this as part of my work on JRebel. We run our test suites on Payara. This involves deploying a quite large quantity of webapps on a server and this issue causes instability in our tests. 

I do understand the issues around re-deployment on Windows. And I would be much less surprised if this issue happened when I tried to redeploy `app-that-causes-crashing`.

However:
1) Is the forceful closure of `JarFiles` really needed during deployment of the application? Perhaps this could be delayed until undeployment?
2) As I understand this, closing the cached `Jarfiles`  fixes an issue that only affects Windows. Considering the side-effects, this seems like a net negative on Unix. If my understanding is correct, then perhaps this could be limited to only Windows?

Please let me know if you would like me to create another issue instead of continuing the discussion here.

--- a/nucleus/admin/util/src/main/java/com/sun/enterprise/admin/util/JarFileUtils.java
+++ b/nucleus/admin/util/src/main/java/com/sun/enterprise/admin/util/JarFileUtils.java
@@ -1,7 +1,7 @@
 /*
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
  *
- * Copyright (c) 2015-2018 Payara Foundation and/or its affiliates. All rights reserved.
+ * Copyright (c) 2015-2021 Payara Foundation and/or its affiliates. All rights reserved.
  *
  * The contents of this file are subject to the terms of either the GNU
  * General Public License Version 2 only ("GPL") or the Common Development
@@ -41,8 +41,11 @@ package com.sun.enterprise.admin.util;
 
 import java.io.IOException;
 import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
 import java.net.URL;
 import java.util.ConcurrentModificationException;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
@@ -80,21 +83,30 @@ public class JarFileUtils {
                 }
             }
             if (files != null && !files.isEmpty()) {
-                Set<JarFile> jarFiles = new HashSet<>();
+                Map<String, JarFile> jarFiles = new HashMap<>();
                 if(jarFileFactoryInstance != null) {
                     synchronized (jarFileFactoryInstance) {
-                        jarFiles.addAll(files.values());
+                        jarFiles.putAll(files);
                     }
                 } else {
-                    jarFiles.addAll(files.values());
+                    jarFiles.putAll(files);
                 }
-                for (JarFile file : jarFiles) {
-                    if (file != null) {
-                        file.close();
+                // Close "file:" JAR files so they can be overwritten on Windows
+                // Leave others open, but remove them from the cache so they
+                // Can be closed when garbage collected.  This prevents
+                // closing JarFile instances opened by URLClassLoader
+                // NOTE: This still may close JarFile instances not opened by the calling process.
+                Method closeMethod = clazz.getMethod("close", JarFile.class);
+                closeMethod.setAccessible(true);
+                for (Map.Entry<String, JarFile> file : jarFiles.entrySet()) {
+                    if (file.getKey().startsWith("file:")) {
+                        file.getValue().close();
+                    } else {
+                        closeMethod.invoke(jarFileFactoryInstance, file.getValue());
                     }
                 }
             }
-        } catch (ClassNotFoundException | IllegalAccessException | SecurityException | IllegalArgumentException | IOException | ConcurrentModificationException ex) {
+        } catch (ClassNotFoundException | IllegalAccessException | SecurityException | IllegalArgumentException | IOException | ConcurrentModificationException | InvocationTargetException | NoSuchMethodException ex) {
             LOG.log(Level.SEVERE, null, ex);
         }
     }

